### PR TITLE
Fix handling of trailing bare CR in str::lines

### LINKED ISSUE #94435: str::lines consumes carriage returns at the end of a line which is undocumented
I tried this code:

```rust
fn main() {
    for line in "a\r".lines() {
        println!("{:?}", line);
    }
}
```

I expected to see this happen: `"a\r"` to be printed.

Instead, this happened: `"a"` is being printed.

At this point fixing this behavior may break actual code (and this would be a quiet change), however documenting this may make sense. This behavior was introduced in Rust 1.4.0 as a result of [implementation of RFC 1212](https://rust-lang.github.io/rfcs/1212-line-endings.html).

Curiously, `BufRead::lines` whose behavior was changed by the same RFC doesn't have this issue. It makes sure that `\r` is followed by `\n` before deleting the carriage return.

### Meta

`rustc --version --verbose`:
```
rustc 1.59.0 (9d1b2106e 2022-02-23)
binary: rustc
commit-hash: 9d1b2106e23b1abd32fce1f17267604a5102f57a
commit-date: 2022-02-23
host: x86_64-unknown-linux-gnu
release: 1.59.0
LLVM version: 13.0.0
```

### PR DESCRIPTION
**Pull request by sugar700**
Continuing from #91191.

Fixes #94435.

### CODE REVIEW

**Comment by thomcc** [comment]
So this is a very clean impl, and I'm in favor of it, as it brings us in line with the documented behavior. This *is* a behavioral change tho, and could break people, I think this needs a FCP, or at least for someone on libs-api to decide it doesn't.

r? @joshtriplett

**Comment by joshtriplett** [comment]
I do think this needs an FCP, yes.

@rust-lang/libs-api, shall we change `lines` to only strip a final trailing CR if followed by an LF?

@rfcbot merge

File: library/core/src/lib.rs [hunk 1]
```diff
@@ -184,6 +184,7 @@
 #![feature(intra_doc_pointers)]
 #![feature(intrinsics)]
 #![feature(lang_items)]
+#![feature(let_else)]
 #![feature(link_llvm_intrinsics)]
 #![feature(macro_metavar_expr)]
 #![feature(min_specialization)]
```
Comment (matthiaskrgr) [inline library/core/src/lib.rs:190]: this has been stabilized and is no longer needed

**Comment by matthiaskrgr** [comment]
@bors r- failed in a rollup (see comment above)
https://github.com/rust-lang/rust/pull/102626

File: library/core/src/lib.rs (hunk 1 above)
Comment (xfix) [reply library/core/src/lib.rs]: Fixed.

**Comment by JohnCSimon** [comment]
Triage:  @xfix 

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

@rustbot ready

**Comment by ChrisDenton** [comment]
This looks good to me.

@bors r+

**Comment by vallentin** [comment]
> This _is_ a behavioral change tho, and could break people

Indeed it did @thomcc. I just spent time tracking down this change. As suddenly one of my released crates, had tests that started failing

**Pull request merged**